SLDS Color Picker

davidlygagnon opened the pull request:
Fixes #1422

### Additional description

---

### CONTRIBUTOR checklist (do not remove)
Please complete for every pull request

* [ ] First-time contributors should sign the Contributor License Agreement. It's a fancy way of saying that you are giving away your contribution to this project. If you haven't before, wait a few minutes and a bot will comment on this pull request with instructions.
* [ ] `npm run lint:fix` has been run and linting passes.
* [ ] Mocha, Jest (Storyshots), and `components/component-docs.json` CI tests pass (`npm test`).
* [ ] Tests have been added for new props to prevent regressions in the future. See [readme](https://github.com/salesforce/design-system-react/blob/master/tests/README.md).
* [ ] Review the appropriate Storybook stories. Open [http://localhost:9001/](http://localhost:9001/).
* [ ] The Accessibility panel of each Storybook story has 0 violations (aXe). Open [http://localhost:9001/](http://localhost:9001/).
* [ ] Review tests are passing in the browser. Open [http://localhost:8001/](http://localhost:8001/).
* [ ] Review markup conforms to [SLDS](https://www.lightningdesignsystem.com/) by looking at [DOM snapshot strings](https://facebook.github.io/jest/docs/en/snapshot-testing.html).

### REVIEWER checklist (do not remove)

* [ ] TravisCI tests pass. This includes linting, Mocha, Jest, Storyshots, and `components/component-docs.json` tests.
* [ ] Tests have been added for new props to prevent regressions in the future. See [readme](https://github.com/salesforce/design-system-react/blob/master/tests/README.md).
* [ ] Review the appropriate Storybook stories. Open [http://localhost:9001/](http://localhost:9001/).
* [ ] The Accessibility panel of each Storybook story has 0 violations (aXe). Open [http://localhost:9001/](http://localhost:9001/).
* [ ] Review tests are passing in the browser. Open [http://localhost:8001/](http://localhost:8001/).
* [ ] Review markup conforms to [SLDS](https://www.lightningdesignsystem.com/) by looking at [DOM snapshot strings](https://facebook.github.io/jest/docs/en/snapshot-testing.html).
###### Required only if there are markup / UX changes
* [ ] Add year-first date and commit SHA to `last-slds-markup-review` in `package.json` and push.
* [ ] Request a review of the deployed Heroku app by the Salesforce UX Accessibility Team.
* [ ] Add year-first review date, and commit SHA, `last-accessibility-review`, to `package.json` and push.
* [ ] While the contributor's branch is checked out, run `npm run local-update` within locally cloned [site repo](https://github.com/salesforce-ux/design-system-react-site) to confirm the site will function correctly at the next release.

davidlygagnon commented:
@interactivellama I fixed a few issues on the color picker mentioned in #1422. 

Let's sync up on what's left to do, thanks.

jamesward commented:
@vargasd I've invited you to the org: https://github.com/salesforce

Once accepted, refresh the CLA bot: https://cla.salesforce.com/status/salesforce/design-system-react/pull/1591

davidlygagnon commented:
@interactivellama 

Color picker tasks after meeting on 8/10/2018

#### - Add missing props:
  - [ ] **className**
    slds-color-picker node
  - [ ] **classNameMenu**
    slds-popover node
  - [ ] **errorText**
    for outer input. has error messaging if present. Do not trigger the error internally, but with a state 					container based on isColorValid being false.
  - [ ] **errorTextMenu**
    for inner input. has error messaging if present within the menu
  - [ ] **hasStaticAlignment**
    menu position locking
  - [ ] **menuPosition**
    all menus have this, just passed down to Dialog
  - [ ] **event.onCancel**
    callback with parameters: event, { trigger: 'cancel' || 'clickOutside' } - This aligns with popover and 				can be the same callback
  - [ ] **events.onChange**
    (already implemented)
    Should return {color: <HEX string>, isValid: <boolean>}
  - [ ] **event.onClose**
    fired after close (See Datepicker)
  - [ ] **event.onOpen**
    fired after open (See Datepicker)
  - [ ] **events.onRequestClose**
    used with isOpen to create a controlled UI (See Datepicker)
  - [ ] **events.onRequestOpen**
    used with isOpen to create a controlled UI (See Datepicker)
  - [ ] **events.onValidateColor**
    Callback that overrides default validator used on outer input. If callback returns false, errorText is shown.
  - [ ] **events.onValidateWorkingColor**
    Callback that overrides default validator used on custom tab menu. If callback returns false, 						errorTextMenu is shown.
  - [ ] **events.onWorkingColorChange**
    (already implemented)
  - [ ] **valueWorking**
    to allow controlling working value - value that is set within the custom tab menu.

#### - Transparent swatch selection:
  - Swatch tab menu should have an option to select transparent.
 -  Transparent is identified as empty string "".
  - Custom tab menu should have a checkbox to select or unselect transparent.
    - If transparent is checked, custom controls (custom color range, hue picker, HEX input, rgb inputs) will be disabled.
    - If transparent is unchecked, custom controls are enabled.
    - When transparent is selected and unchecked (default state), custom controls will be set to #000000, R:0, G:0, B:0.
    - This will fix the current issue where custom tab menu is in a weird state when transparent is selected.

#### - Fix following bugs:
 - where outer input changes as working input value changes. Outer input should only change when clicking “Done” button.
			
#### - Snapshot tests

##### Notes:

We discussed the necessity of having both events.onValidateColor and events.onValidateWorkingColor 				One could argue that two different validation functions could be passed in. We came up to the 				conclusion that having both functions add clarify and helps differentiate which input is validation being triggered on (either outer input or input in the custom tab menu). Validation and error message happens in both case and it’s possible that the outer input text is hidden (by using inputHidden prop), therefore requiring the need of both.

davidlygagnon commented:
@vargasd, can you accept the CLA as @jamesward mentioned? I think that's why this PR is failing the CLA check.

interactivellama commented:
### Feedback from Alan W:
Looks good to me. Thanks for exploring this. One quick note on the visual design side: Please make the red slash 1 or 2px wide. It's a little aggressive currently. I also suggest using a red that isn't so close to an error. Maybe use MAROON or something similar. 

This is SLDS Maroon:
rgb(135, 5, 0)
#870500
MAROON

davidlygagnon commented on components/color-picker/index.jsx:
```diff
@@ -0,0 +1,582 @@
+			this.props.events.onValidateWorkingColor
+		);
+		this.state = {
+			currentColor: this.props.value || this.props.swatchColors[0],
```
Note here - if no value passed in, we'll use the first color in the set for the current color, instead of defaulting to transparent.

davidlygagnon replied:
It just felt strange to set the current color to "transparent" by default - which is not even in the set of pre-defined colors. By selecting the first one, when the popover opens, the top right color is focused, which feels right as a starting point. I also assume that most consumer won't be selecting transparent as their main color but that's just an assumption.

davidlygagnon commented on components/color-picker/index.jsx:
```diff
@@ -0,0 +1,582 @@
+				)}
+				disabled={this.props.disabled}
+				id={`color-picker-summary-input-${this.generatedId}`}
+				maxLength="7"
```
Just something to be aware of is, if we would want to support 8 digits hex, where the two last digits are for opacity, we would have to change this `maxLength`.

interactivellama replied:
SLDS doesn't have a max length. I'm OK with removing it.

davidlygagnon commented on components/radio-button-group/index.jsx:
```diff
@@ -108,21 +115,31 @@ class RadioButtonGroup extends React.Component {
 					)}
 				>
-					<div className="slds-radio_button-group">{children}</div>
+					<div style={this.props.style} className="slds-radio_button-group">
```
Note: undocumented style prop needed to remove some unnecessary styling.

interactivellama commented:
Deployed to https://design-system-react-co-pr-1591.herokuapp.com/?selectedKind=SLDSColorPicker&selectedStory=Default&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybook%2Factions%2Factions-panel in case you want to show your team/designer, etc.

interactivellama commented:
@davidlygagnon 

- [ ] The swatch panel should be navigable with arrow keys. See Expected Keyboard Interactions on https://www.lightningdesignsystem.com/components/color-picker/ The code would be similar to Datepicker or Combobox menu item up and down. 

> This element has the role of listbox, and keyboard interactions when its <a> children are focused needs to behave in a menu-like fashion. Keypresses need to move the actively focused element in the appropriate direction (left/up will move to the previous item, down/right will move to the next item).

> Additionally, when focused on the first or last item, the appropriate key action needs to move the focus to the last or first item, respectively. It is expected to behave in a cyclical fashion.

interactivellama commented:
- [ ] Clicking button/opening crash component in this example: https://design-system-react-co-pr-1591.herokuapp.com/?selectedKind=SLDSColorPicker&selectedStory=Outer%20Input%20in%20Error%20State&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybook%2Factions%2Factions-panel

![screen shot 2018-10-18 at 4 34 51 pm](https://user-images.githubusercontent.com/1290832/47188087-d3607500-d2f3-11e8-8a46-737603b7f6cf.png)

interactivellama commented:
TravisCI is [currently complaining](https://travis-ci.org/salesforce/design-system-react/builds/443413564#L6775) about an empty comment description for a propType. Please look through the propTypes and make sure they all have comments before them.

interactivellama commented on components/color-picker/index.jsx:
```diff
@@ -0,0 +1,585 @@
+			<div
+				className={classNames('slds-color-picker', this.props.className)}
+				ref={(node) => {
+					this.wrapper = node;
```
What is this used for?

interactivellama commented on package.json:
```diff
@@ -341,6 +341,18 @@
 			"SLDS-component-path": "/components/checkbox",
 			"url-slug": "checkboxes"
 		},
+		{
```
Why two?

Also this needs a `url-slug`

davidlygagnon commented on components/color-picker/index.jsx:
```diff
@@ -0,0 +1,585 @@
+	 * * `submitButton`: Text for submit/done button of popover
+	 */
+	labels: PropTypes.shape({
+		blueAbbreviated: PropTypes.string,
```
Looks like Sam did it but forgot to expose the prop. I'll fix it.

davidlygagnon commented:
@interactivellama I addressed your comments. Please take a look, thanks!

Here's a preview of navigating with keyboard (left/up goes to previous color, right/down goes to next color).

![color-picker-keyboard](https://user-images.githubusercontent.com/1821787/47235253-54b81600-d38d-11e8-8939-8e15ceacaa2c.gif)

davidlygagnon commented:
@garygong and I were discussing about the selected color border style in the pre-defined set of colors, which is what I came up with:

![screen shot 2018-10-19 at 1 40 56 pm](https://user-images.githubusercontent.com/1821787/47242755-a3bd7580-d3a4-11e8-9c6b-a5b278ff9b9a.png)

There's an SLDS version, show below, but it's not very easy to see which color is selected. Are there any recommendations on styles to use?

SLDS version of focused:
![screen shot 2018-10-19 at 1 42 22 pm](https://user-images.githubusercontent.com/1821787/47242816-d5ced780-d3a4-11e8-860e-9e582ed6a070.png)
![screen shot 2018-10-19 at 1 42 55 pm](https://user-images.githubusercontent.com/1821787/47242833-e97a3e00-d3a4-11e8-9296-0f0ffce0b558.png)

interactivellama commented:
I found a few issues from console errors in storybook. I've gone ahead and pushed them up. I'll push them over to .8 also

### Adds:
* Removes default label of “Choose Color” to allow assistiveText.label in the future
* Add label comment for prop description
* Fix defaultSelectedTab propType parameters
* Pass merged labels object around from render funtion to allow partial objects
* Update input import
